core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1877,6 +1877,35 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
    */
   final def untraced: ZIO[R, E, A] = tracingStatus(TracingStatus.Untraced)
 
+  /**
+   * Sequentially zips the this result with the specified result. Combines both
+   * `Cause[E1]` when both both effects fail.
+   */
+  final def validate[R1 <: R, E1 >: E, B](that: ZIO[R1, E1, B]): ZIO[R1, E1, (A, B)] =
+    validateWith(that)((_, _))
+
+  /**
+   * Returns an effect that executes both this effect and the specified effect,
+   * in parallel. Combines both Cause[E1]` when both both effects fail.
+   */
+  final def validatePar[R1 <: R, E1 >: E, B](that: ZIO[R1, E1, B]): ZIO[R1, E1, (A, B)] =
+    validateWithPar(that)((_, _))
+
+  /**
+   * Sequentially zips this effect with the specified effect using the
+   * specified combiner function. Combines the causes in case both effect fail.
+   */
+  final def validateWith[R1 <: R, E1 >: E, B, C](that: ZIO[R1, E1, B])(f: (A, B) => C): ZIO[R1, E1, C] =
+    self.run.zipWith(that.run)(_ zip _).flatMap(ZIO.done(_)).map(f.tupled)
```

Comment:
We can save a couple of allocations here by pattern matching on the exit values in `zipWith` / `zipWithPar` and combining them into the desired type instead of doing the subsequent `flatMap` and `map`. We should be sure to combine the failures using `Cause.Then` in the sequential case and `Cause.Both` in the parallel case.

Reply:
```scala
final def validateWith[R1 <: R, E1 >: E, B, C](that: ZIO[R1, E1, B])(f: (A, B) => C): ZIO[R1, E1, C] =
    self.run.flatMap { exit =>
      that.run.flatMap { (exit, _) match {
        case (Exit.Success(a), Exit.Success(b)) => ZIO.succeedNow(f(a, b))
        case (Exit.Failure(l), Exit.Failure(r)) => ZIO.halt(Cause.Then(l, r))
        case (Exit.Failure(cause), _)           => ZIO.halt(cause)
        case (_, Exit.Failure(cause))           => ZIO.halt(cause)
      }}
    }

  final def validateWithPar[R1 <: R, E1 >: E, B, C](that: ZIO[R1, E1, B])(f: (A, B) => C): ZIO[R1, E1, C] =
    self.run.zipWithPar(that.run) {
      case (Exit.Success(a), Exit.Success(b)) => ZIO.succeedNow(f(a, b))
      case (Exit.Failure(l), Exit.Failure(r)) => ZIO.halt(Cause.Both(l, r))
      case (Exit.Failure(cause), _)           => ZIO.halt(cause)
      case (_, Exit.Failure(cause))           => ZIO.halt(cause)
    }.flatten
```
@adamgfraser like this?

Reply:
Yes though I think we can clean it up even more if we use `zipWith` on `Exit`. So we could do:

```scala
self.run.zipWith(that.run)(_.zipWith(_)(f, _ ++ _)).map(ZIO.done(_))
self.run.zipWithPar(that.run)(_.zipWith(_)(f, _ && _)).map(ZIO.done(_))
```

Reply:
Right, much better. Done. @adamgfraser